File: learning-platform/backend-python/app/routers/learning.py
from fastapi import APIRouter, Depends, File, HTTPException, UploadFile
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.schemas.common import ok
from app.services import learning_service as service


router = APIRouter(tags=["learning-platform"])


class KnowledgeBaseCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=500)


class ProgressUpdate(BaseModel):
    status: str


class GenerateQuestionsRequest(BaseModel):
    difficulty: str = "medium"
    count: int = Field(default=5, ge=5, le=20)


class QuizAnswer(BaseModel):
    questionId: int
    userAnswer: str


class SubmitQuizRequest(BaseModel):
    answers: list[QuizAnswer]
    durationSeconds: int = 0


def _handle(action):
    try:
        return ok(action())
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/knowledge-bases")
def list_knowledge_bases(db: Session = Depends(get_db)) -> dict:
    return ok(service.list_knowledge_bases(db))


@router.post("/knowledge-bases")
def create_knowledge_base(payload: KnowledgeBaseCreate, db: Session = Depends(get_db)) -> dict:
    return _handle(lambda: service.create_knowledge_base(db, payload.name, payload.description))


@router.delete("/knowledge-bases/{kb_id}")
def delete_knowledge_base(kb_id: int, db: Session = Depends(get_db)) -> dict:
    return _handle(lambda: service.delete_knowledge_base(db, kb_id))


@router.get("/knowledge-bases/{kb_id}/documents")
def list_documents(kb_id: int, db: Session = Depends(get_db)) -> dict:
    return _handle(lambda: service.list_documents(db, kb_id))


@router.post("/knowledge-bases/{kb_id}/documents")
def upload_document(kb_id: int, file: UploadFile = File(...), db: Session = Depends(get_db)) -> dict:
    return _handle(lambda: service.save_uploaded_document(db, kb_id, file))


@router.delete("/documents/{document_id}")
def delete_document(document_id: int, db: Session = Depends(get_db)) -> dict:
    return _handle(lambda: service.delete_document(db, document_id))


@router.get("/knowledge-bases/{kb_id}/search")
def search_chunks(kb_id: int, keyword: str, db: Session = Depends(get_db)) -> dict:
    return _handle(lambda: service.search_chunks(db, kb_id, keyword))


@router.post("/knowledge-bases/{kb_id}/outline")
def generate_outline(kb_id: int, db: Session = Depends(get_db)) -> dict:
    return _handle(lambda: service.generate_outline(db, kb_id))


@router.get("/knowledge-bases/{kb_id}/chapters")
def list_chapters(kb_id: int, db: Session = Depends(get_db)) -> dict:
    return _handle(lambda: service.list_chapters(db, kb_id))


@router.get("/chapters/{chapter_id}/content")
def chapter_content(chapter_id: int, db: Session = Depends(get_db)) -> dict:
    return _handle(lambda: service.get_chapter_content(db, chapter_id))


@router.patch("/chapters/{chapter_id}/progress")
def update_progress(chapter_id: int, payload: ProgressUpdate, db: Session = Depends(get_db)) -> dict:
    return _handle(lambda: service.set_progress(db, chapter_id, payload.status))


@router.get("/knowledge-bases/{kb_id}/progress")
def progress_overview(kb_id: int, db: Session = Depends(get_db)) -> dict:
    return _handle(lambda: service.progress_overview(db, kb_id))


@router.delete("/knowledge-bases/{kb_id}/progress")
def reset_progress(kb_id: int, db: Session = Depends(get_db)) -> dict:
    return _handle(lambda: service.reset_progress(db, kb_id))


@router.post("/chapters/{chapter_id}/questions")
def generate_questions(chapter_id: int, payload: GenerateQuestionsRequest, db: Session = Depends(get_db)) -> dict:
    return _handle(lambda: service.generate_questions(db, chapter_id, payload.difficulty, payload.count))


@router.get("/chapters/{chapter_id}/questions")
def list_questions(chapter_id: int, db: Session = Depends(get_db)) -> dict:
    return _handle(lambda: service.list_questions(db, chapter_id))


@router.post("/chapters/{chapter_id}/quiz")
def submit_quiz(chapter_id: int, payload: SubmitQuizRequest, db: Session = Depends(get_db)) -> dict:
    answers = [answer.model_dump() for answer in payload.answers]
    return _handle(lambda: service.submit_quiz(db, chapter_id, answers, payload.durationSeconds))


@router.get("/quiz/history")
def quiz_history(db: Session = Depends(get_db)) -> dict:
    return ok(service.quiz_history(db))


@router.get("/wrong-questions")
def wrong_questions(db: Session = Depends(get_db)) -> dict:
    return ok(service.wrong_question_list(db))

File: learning-platform/frontend/src/api/learning.ts
interface LearningApiResponse<T> {
  code: number;
  message: string;
  data: T;
}

export interface KnowledgeBase {
  id: number;
  name: string;
  description?: string;
  documentCount: number;
  chunkCount: number;
  vectorCollection: string;
  createTime: string;
  updateTime: string;
}

export interface KnowledgeDocument {
  id: number;
  knowledgeBaseId: number;
  filename: string;
  fileType: string;
  parserStatus: string;
  chunkCount: number;
  errorMessage?: string;
  createTime: string;
}

export interface Chapter {
  id: number;
  knowledgeBaseId: number;
  parentId: number;
  title: string;
  summary: string;
  sortNo: number;
}

export interface ChapterContent extends Chapter {
  content: Array<{ title: string; text: string }>;
}

export interface ProgressOverview {
  total: number;
  completed: number;
  percent: number;
}

export interface Question {
  id: number;
  chapterId: number;
  questionType: string;
  difficulty: string;
  stem: string;
  options: string[];
  answer: string;
  analysis: string;
}

export interface QuizResult {
  attemptId: number;
  score: number;
  total: number;
  correct: number;
  details: Array<{
    question: Question;
    userAnswer: string;
    isCorrect: boolean;
  }>;
}

export interface QuizHistory {
  id: number;
  chapterId: number;
  chapterTitle: string;
  score: number;
  durationSeconds: number;
  totalCount: number;
  correctCount: number;
  createTime: string;
}

export interface WrongQuestion {
  id: number;
  questionId: number;
  chapterId: number;
  chapterTitle: string;
  difficulty: string;
  stem: string;
  options: string[];
  answer: string;
  analysis: string;
  wrongCount: number;
  correctCount: number;
  accuracy: number;
  lastWrongTime: string;
}

export interface SearchChunk {
  id: number;
  documentId: number;
  chunkIndex: number;
  preview: string;
  score?: number;
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`/api${path}`, init);
  const result = (await response.json()) as LearningApiResponse<T>;
  if (!response.ok || result.code !== 200) {
    throw new Error(result.message || '学习平台接口请求失败');
  }
  return result.data;
}

export function fetchKnowledgeBases(): Promise<KnowledgeBase[]> {
  return request<KnowledgeBase[]>('/knowledge-bases');
}

export function createKnowledgeBase(payload: { name: string; description?: string }): Promise<KnowledgeBase> {
  return request<KnowledgeBase>('/knowledge-bases', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
}

export function deleteKnowledgeBase(id: number): Promise<void> {
  return request<void>(`/knowledge-bases/${id}`, { method: 'DELETE' });
}

export function fetchDocuments(kbId: number): Promise<KnowledgeDocument[]> {
  return request<KnowledgeDocument[]>(`/knowledge-bases/${kbId}/documents`);
}

export function uploadDocument(kbId: number, file: File): Promise<KnowledgeDocument> {
  const body = new FormData();
  body.append('file', file);
  return request<KnowledgeDocument>(`/knowledge-bases/${kbId}/documents`, {
    method: 'POST',
    body,
  });
}

export function deleteDocument(id: number): Promise<void> {
  return request<void>(`/documents/${id}`, { method: 'DELETE' });
}

export function searchKnowledge(kbId: number, keyword: string): Promise<SearchChunk[]> {
  return request<SearchChunk[]>(`/knowledge-bases/${kbId}/search?keyword=${encodeURIComponent(keyword)}`);
}

export function generateOutline(kbId: number): Promise<Chapter[]> {
  return request<Chapter[]>(`/knowledge-bases/${kbId}/outline`, { method: 'POST' });
}

export function fetchChapters(kbId: number): Promise<Chapter[]> {
  return request<Chapter[]>(`/knowledge-bases/${kbId}/chapters`);
}

export function fetchChapterContent(chapterId: number): Promise<ChapterContent> {
  return request<ChapterContent>(`/chapters/${chapterId}/content`);
}

export function updateChapterProgress(chapterId: number, status: string): Promise<void> {
  return request<void>(`/chapters/${chapterId}/progress`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ status }),
  });
}

export function fetchProgressOverview(kbId: number): Promise<ProgressOverview> {
  return request<ProgressOverview>(`/knowledge-bases/${kbId}/progress`);
}

export function resetProgress(kbId: number): Promise<void> {
  return request<void>(`/knowledge-bases/${kbId}/progress`, { method: 'DELETE' });
}

export function generateQuestions(chapterId: number, difficulty: string, count: number): Promise<Question[]> {
  return request<Question[]>(`/chapters/${chapterId}/questions`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ difficulty, count }),
  });
}

export function fetchQuestions(chapterId: number): Promise<Question[]> {
  return request<Question[]>(`/chapters/${chapterId}/questions`);
}

export function submitQuiz(
  chapterId: number,
  answers: Array<{ questionId: number; userAnswer: string }>,
  durationSeconds: number,
): Promise<QuizResult> {
  return request<QuizResult>(`/chapters/${chapterId}/quiz`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ answers, durationSeconds }),
  });
}

export function fetchQuizHistory(): Promise<QuizHistory[]> {
  return request<QuizHistory[]>('/quiz/history');
}

export function fetchWrongQuestions(): Promise<WrongQuestion[]> {
  return request<WrongQuestion[]>('/wrong-questions');
}
